pkg/cmd/cli/cmd/set/buildsecret.go

```diff
@@ -0,0 +1,283 @@
+			kcmdutil.CheckErr(options.Validate())
+			if err := options.Run(); err != nil {
+				// TODO: move me to kcmdutil
+				if err == cmdutil.ErrExit {
+					os.Exit(1)
+				}
+				kcmdutil.CheckErr(err)
+			}
+		},
+	}
+
+	kcmdutil.AddPrinterFlags(cmd)
+	cmd.Flags().StringVarP(&options.Selector, "selector", "l", options.Selector, "Selector (label query) to filter build configs")
+	cmd.Flags().BoolVar(&options.All, "all", options.All, "Select all build configs in the namespace")
+	cmd.Flags().StringSliceVarP(&options.Filenames, "filename", "f", options.Filenames, "Filename, directory, or URL to file to use to edit the resource.")
+
+	cmd.Flags().BoolVar(&options.Push, "push", options.Push, "If true, set the push secret on a build config")
+	cmd.Flags().BoolVar(&options.Pull, "pull", options.Pull, "If true, set the pull secret on a build config")
+	cmd.Flags().BoolVar(&options.Source, "source", options.Source, "If true, set the source secret on a build config")
+	cmd.Flags().BoolVar(&options.Remove, "remove", options.Remove, "If true, remove the build secret.")
+
+	cmd.MarkFlagFilename("filename", "yaml", "yml", "json")
+
+	return cmd
+}
+
+func (o *BuildSecretOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, args []string) error {
+	if !o.Remove {
+		if len(args) < 1 {
+			return kcmdutil.UsageError(cmd, "a secret name must be specified")
+		}
+		o.Secret = args[len(args)-1]
+		args = args[:len(args)-1]
+	}
+	resources := args
+	if len(resources) == 0 && len(o.Selector) == 0 && len(o.Filenames) == 0 && !o.All {
+		return kcmdutil.UsageError(cmd, "one or more build configs must be specified as <name> or <resource>/<name>")
+	}
+
+	cmdNamespace, explicit, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+	clientConfig, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+	o.OutputVersion, err = kcmdutil.OutputVersion(cmd, clientConfig.GroupVersion)
+	if err != nil {
+		return err
+	}
+
+	mapper, typer := f.Object(false)
+	o.Builder = resource.NewBuilder(mapper, typer, resource.ClientMapperFunc(f.ClientForMapping), kapi.Codecs.UniversalDecoder()).
+		ContinueOnError().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(explicit, false, o.Filenames...).
+		SelectorParam(o.Selector).
+		ResourceNames("buildconfigs", resources...).
+		Flatten()
+
+	if o.All {
+		o.Builder.ResourceTypes("buildconfigs").SelectAllParam(o.All)
+	}
+
+	output := kcmdutil.GetFlagString(cmd, "output")
+	if len(output) != 0 {
+		o.PrintObject = func(obj runtime.Object) error { return f.PrintObject(cmd, mapper, obj, o.Out) }
+	}
+
+	o.Encoder = f.JSONEncoder()
+	o.ShortOutput = kcmdutil.GetFlagString(cmd, "output") == "name"
+	o.Mapper = mapper
+
+	return nil
+}
+
+func (o *BuildSecretOptions) Validate() error {
+	if !o.Pull && !o.Push && !o.Source {
+		return fmt.Errorf("specify the type of secret to set (--push, --pull, or --source)")
+	}
+	if !o.Remove && len(o.Secret) == 0 {
```

Comment:
If I specify `--remove` and a secret name, my expectation is that you'll remove that secret from all categories.  Doesn't break existing contracts, so it could come later.